File: bindings/java/src/main/java/dev/tachyon_ipc/RxBatchGuard.java
package dev.tachyon_ipc;

import java.lang.foreign.Arena;
import java.lang.foreign.MemoryLayout;
import java.lang.foreign.MemorySegment;
import java.util.Iterator;
import java.util.NoSuchElementException;
import java.util.Objects;
import java.util.Spliterator;
import java.util.Spliterators;
import java.util.function.Consumer;

/**
 * A high-throughput batch lease that drains multiple contiguous messages in a single FFI crossing.
 *
 * @implSpec Backed by a confined FFM Arena holding the native C-struct array. Committing the batch
 * advances the consumer head for all processed messages simultaneously.
 */
public final class RxBatchGuard implements AutoCloseable, Iterable<RxMsgView> {

	/**
	 * The native bus pointer used for FFI operations.
	 */
	private final MemorySegment busHandle;

	/**
	 * The confined memory arena governing the lifecycle of the native struct array and its projections.
	 */
	private final Arena arena;

	/**
	 * The absolute number of messages successfully extracted from the ring buffer.
	 */
	private final long count;

	/**
	 * The array of object-oriented views mapping to the native memory segments.
	 */
	private final RxMsgView[] views;

	/**
	 * The raw FFM memory segment representing the contiguous array of {@code tachyon_msg_view_t} structs.
	 */
	private final MemorySegment viewsArray;

	/**
	 * State flag indicating whether the batch has been acknowledged and the consumer head advanced.
	 */
	private boolean consumed;

	/**
	 * Internal constructor invoked exclusively by the {@link TachyonBus}.
	 * Allocates the layout and invokes the native ABI to populate the structs.
	 *
	 * @param busHandle     The native bus pointer.
	 * @param busArena      The shared arena of the bus used to reinterpret raw pointers.
	 * @param maxMsgs       The upper limit of messages to extract.
	 * @param spinThreshold The CPU yield threshold before blocking.
	 * @throws IllegalStateException If the native ABI returns a count exceeding the requested bounds.
	 */
	RxBatchGuard(MemorySegment busHandle, Arena busArena, int maxMsgs, int spinThreshold) {
		this.busHandle = busHandle;
		this.consumed = false;
		this.arena = Arena.ofConfined();

		try {
			MemoryLayout arrayLayout = MemoryLayout.sequenceLayout(maxMsgs, MsgViewLayout.layout);
			this.viewsArray = arena.allocate(arrayLayout);
			this.count = TachyonABI.drainBatch(busHandle, viewsArray, maxMsgs, spinThreshold);

			if (this.count < 0 || this.count > maxMsgs) {
				throw new IllegalStateException("Invalid batch count returned from C ABI: " + this.count);
			}

			if (this.count == 0) {
				this.views = new RxMsgView[0];
			} else {
				this.views = new RxMsgView[(int) this.count];
				for (int i = 0; i < this.count; i++) {
					long offset = i * MsgViewLayout.sizeBytes;
					MemorySegment viewStruct = viewsArray.asSlice(offset, MsgViewLayout.sizeBytes);
					MemorySegment ptr = (MemorySegment) MsgViewLayout.ptrHandle.get(viewStruct);
					long actualSize = (long) MsgViewLayout.sizeHandle.get(viewStruct);
					int typeId = (int) MsgViewLayout.typeHandle.get(viewStruct);
					MemorySegment safePtr = ptr.reinterpret(actualSize, busArena, null);
					this.views[i] = new RxMsgView(safePtr, typeId, actualSize);
				}
			}
		} catch (Throwable throwable) {
			arena.close();
			throw throwable;
		}
	}

	/**
	 * Retrieves the number of valid messages loaded into this batch.
	 *
	 * @return The total message count.
	 */
	public long getCount() {
		return count;
	}

	/**
	 * Retrieves a specific message view from the batch by its index.
	 *
	 * @param index The zero-based index of the message within the batch.
	 * @return The corresponding {@link RxMsgView}.
	 * @throws IllegalStateException          If the batch has already been committed or closed.
	 * @throws ArrayIndexOutOfBoundsException If the index is out of bounds.
	 */
	public RxMsgView get(int index) {
		checkState();
		return views[index];
	}

	/**
	 * Submits the entire batch of messages to the native arena, advancing the consumer head
	 * in a single atomic operation.
	 *
	 * @implSpec Iterates through all instantiated {@link RxMsgView} references and invalidates
	 * their memory segments to strictly enforce zero-copy lifetime constraints and prevent use-after-free.
	 * Finally, closes the confined struct arena.
	 */
	public void commit() {
		checkState();
		this.consumed = true;

		try {
			if (count > 0) {
				TachyonABI.commitRxBatch(busHandle, viewsArray, count);
			}
		} finally {
			for (RxMsgView view : views) {
				view.invalidate();
			}

			arena.close();
		}
	}

	/**
	 * Finalizes the batch lifecycle.
	 *
	 * @implSpec Automatically commits the batch if not explicitly consumed, ensuring the
	 * consumer head does not stall the producer arena.
	 */
	@Override
	public void close() {
		if (!consumed) {
			commit();
		}
	}

	/**
	 * Provides an iterator over the valid messages in this batch.
	 *
	 * @return An {@link Iterator} of {@link RxMsgView}.
	 * @throws IllegalStateException If the batch has already been committed.
	 */
	@Override
	public Iterator<RxMsgView> iterator() {
		checkState();
		return new Iterator<>() {
			private int currentIndex = 0;

			@Override
			public boolean hasNext() {
				checkState();
				return currentIndex < views.length;
			}

			@Override
			public RxMsgView next() {
				if (!hasNext()) {
					throw new NoSuchElementException();
				}

				return views[currentIndex++];
			}
		};
	}

	/**
	 * Performs the given action for each element of the batch until all elements have been processed.
	 *
	 * @param action The action to be performed for each message view.
	 * @throws NullPointerException  If the specified action is null.
	 * @throws IllegalStateException If the batch is committed during iteration.
	 */
	@Override
	public void forEach(Consumer<? super RxMsgView> action) {
		Objects.requireNonNull(action);
		for (RxMsgView view : views) {
			checkState();
			action.accept(view);
		}
	}

	/**
	 * Creates a Spliterator over the message views in this batch.
	 *
	 * @return A {@link Spliterator} with ORDERED and IMMUTABLE characteristics.
	 * @throws IllegalStateException If the batch has already been committed.
	 */
	@Override
	public Spliterator<RxMsgView> spliterator() {
		checkState();
		return Spliterators.spliterator(
				views,
				Spliterator.ORDERED | Spliterator.SIZED | Spliterator.SUBSIZED | Spliterator.NONNULL | Spliterator.IMMUTABLE
		);
	}

	/**
	 * Validates the temporal lifecycle of the batch guard.
	 *
	 * @throws IllegalStateException If the guard has already been committed or closed.
	 */
	private void checkState() {
		if (consumed || !arena.scope().isAlive()) {
			throw new IllegalStateException("RxBatchGuard has already been committed or closed");
		}
	}
}

File: bindings/csharp/src/TachyonIpc/RxBatchGuard.cs
using TachyonIpc.Native;

namespace TachyonIpc;

/// <summary>
/// Lightweight read-only view over a single message inside an <see cref="RxBatchGuard"/>.
/// Valid only within the parent guard's scope.
/// </summary>
public unsafe ref struct MsgView
{
    private readonly byte* _ptr;
    private readonly nuint _actualSize;

    /// <summary>bits [15:0] = msg_type, bits [31:16] = route_id.</summary>
    public readonly uint TypeId;

    public ushort MsgType => TachyonIpc.TypeId.MsgType(TypeId);
    public ushort RouteId => TachyonIpc.TypeId.RouteId(TypeId);

    /// <summary>
    /// Read-only view of the message payload.
    /// Do not retain beyond the parent <see cref="RxBatchGuard"/> scope.
    /// </summary>
    public ReadOnlySpan<byte> Data => new(_ptr, checked((int)_actualSize));

    internal MsgView(void* ptr, nuint actualSize, uint typeId)
    {
        _ptr = (byte*)ptr;
        _actualSize = actualSize;
        TypeId = typeId;
    }
}

/// <summary>
/// Read-side guard for a batch of messages.
/// Returned by Bus.TryReceiveBatch / Bus.DrainBatch.
/// Must be used with <c>using</c>.
/// Dispose() commits all slots at once.
/// </summary>
/// <remarks>
/// ref struct, stack-only, zero allocation.
/// All message pointers are valid only until Dispose() is called.
/// </remarks>
public unsafe ref struct RxBatchGuard
{
    private readonly nint _bus;
    private readonly TachyonMsgView* _views;
    private readonly nuint _count;
    private bool _done;

    /// <summary>Number of messages received in this batch.</summary>
    public int Count => (int)_count;

    /// <summary>True if the batch is empty.</summary>
    public bool IsEmpty => _count == 0;

    internal RxBatchGuard(nint bus, TachyonMsgView* views, nuint count)
    {
        _bus = bus;
        _views = views;
        _count = count;
        _done = false;
    }

    /// <summary>Returns the message at <paramref name="index"/>.</summary>
    public MsgView this[int index]
    {
        get
        {
            if (_done)
                ThrowInvalidated();
            if ((uint)index >= _count)
                throw new ArgumentOutOfRangeException(nameof(index));
            ref var v = ref _views[index];
            return new MsgView(v.Ptr, v.ActualSize, v.TypeId);
        }
    }

    /// <summary>Commits all slots in the batch. Called automatically by Dispose.</summary>
    public void Commit()
    {
        if (_done)
            return;
        if (_count == 0)
            return;
        TachyonException.ThrowIfError(
            TachyonNative.tachyon_commit_rx_batch(_bus, _views, _count), nameof(TachyonNative.tachyon_commit_rx_batch)
        );
        _done = true;
    }

    /// <summary>Commits all slots and releases the batch.</summary>
    public void Dispose() => Commit();

    /// <summary>Supports <c>foreach</c> iteration without allocation.</summary>
    public Enumerator GetEnumerator() => new(this);

    public ref struct Enumerator
    {
        private readonly RxBatchGuard _guard;
        private int _index;

        internal Enumerator(RxBatchGuard guard)
        {
            _guard = guard;
            _index = -1;
        }

        public bool MoveNext() => ++_index < _guard.Count;

        public MsgView Current => _guard[_index];
    }

    private static void ThrowInvalidated() =>
        throw new InvalidOperationException("RxBatchGuard has already been committed.");
}